Debugging iOS release applications using lldb

su-vikas opened the pull request:
For issue #1371 

Added debugging using lldb, and a remark on use of gdb.

sushi2k commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
````diff
@@ -404,6 +404,60 @@ for armv7.
 Attaching to process 2670...
 ```
 
+###### Launching an application via debugserver
````
Instead of going so deep with the titles (this is now going 6 levels deep), just make a sentence instead:

With the following command you can launch an application via debugserver:

cpholguera replied:
```suggestion
With the following command you can launch an application via debugserver:
```

sushi2k commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
````diff
@@ -404,6 +404,60 @@ for armv7.
+(lldb) process connect connect://iOSIP:1234
+```
+
+- `image list` gives a list of main executable and all dependent libraries.
````
No need to make a bullet point here, just remove the `-`

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+In the previous section we learned about how to setup a debugging environment on an iOS device using `lldb`. In this section we will use this information and learn how to debug a 3rd party release application. We will continue using the [UnCrackable Level 1 crackme app](https://github.com/OWASP/owasp-mstg/blob/master/Crackmes/iOS/Level_01/UnCrackable_Level1.ipa "UnCrackable Level 1 iOS App") and  ...
+
+What differentiates a debug build of an application from a release build? In a release build, the compiled code is optimized to achieve maximum performance and minimum binary build size. As a general practice, in a release build all the debug symbols are usually stripped to reduce binary size and also to prevent reverse engineering of the binaries. Due to absence of these debug symbols, setting o ...
```
Simplified and split.

```suggestion
In contrast to a debug build, the code compiled for a release build is optimized to achieve maximum performance and minimum binary build size. As a general best practice, most of the debug symbols are stripped for a release build, adding a layer of complexity when reverse engineering and debugging the binaries.

Due to the absence of the debug symbols, symbol names are missing from the backtrace outputs and setting breakpoints by simply using function names is not possible. Fortunately, debuggers also support setting breakpoints directly on memory addresses. Further in this section we will learn how to do so and eventually solve the crackme challenge.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+What differentiates a debug build of an application from a release build? In a release build, the compiled code is optimized to achieve maximum performance and minimum binary build size. As a general practice, in a release build all the debug symbols are usually stripped to reduce binary size and also to prevent reverse engineering of the binaries. Due to absence of these debug symbols, setting o ...
+
+Some groundwork is needed before setting a breakpoint using memory addresses. It requires determining two addresses:
```
Added new line and changed address to offsets
```suggestion
Some groundwork is needed before setting a breakpoint using memory addresses. It requires determining two offsets:

```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+1. Breakpoint offset: The _address offset_ of the code where we want to set a breakpoint. This address is obtained by performing static analysis of the code in a disassembler like Ghidra.
+2. ASLR shift offset: The _ASLR shift offset_ for the current process. Since ASLR offset is randomly generated on every new instance of an application, this has to be obtained for every debugging session individually. This is determined using the debugger itself.
+
+The final breakpoint address to be used in the debugger is the sum of the above two addresses (Breakpoint offset + ASLR shift offset). This approach assumes that the image base address (discussed shortly) used by Ghidra and iOS is the same, which is true most of the times.
```
```suggestion
The final breakpoint address to be used in the debugger is the sum of the above two addresses (Breakpoint offset + ASLR shift offset). This approach assumes that the image base address (discussed shortly) used by the disassembler and iOS is the same, which is true most of the time.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+The final breakpoint address to be used in the debugger is the sum of the above two addresses (Breakpoint offset + ASLR shift offset). This approach assumes that the image base address (discussed shortly) used by Ghidra and iOS is the same, which is true most of the times.
+
+When a binary is opened in a tool like Ghidra, it loads a binary by emulating the respective operating system's loader. The address at which the binary is loaded is called _image base address_. All the code and symbols inside this binary can be addressed using a constant address offset from this image base address. In Ghidra, the image base address can be obtained by determining the address of th ...
```
```suggestion
When a binary is opened in a disassembler like Ghidra, it loads a binary by emulating the respective operating system's loader. The address at which the binary is loaded is called _image base address_. All the code and symbols inside this binary can be addressed using a constant address offset from this image base address. In Ghidra, the image base address can be obtained by determining the address of the start of a Mach-O file. In this case, it is 0x100000000.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+When a binary is opened in a tool like Ghidra, it loads a binary by emulating the respective operating system's loader. The address at which the binary is loaded is called _image base address_. All the code and symbols inside this binary can be addressed using a constant address offset from this image base address. In Ghidra, the image base address can be obtained by determining the address of th ...
+
+![Obtain image base address using Ghidra](Images/Chapters/0x06c/debugging_ghidra_image_base_address.png "Obtain image base address using Ghidra")
```
```suggestion
![Obtaining image base address using Ghidra](Images/Chapters/0x06c/debugging_ghidra_image_base_address.png "Obtaining image base address using Ghidra")
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+![Obtain image base address using Ghidra](Images/Chapters/0x06c/debugging_ghidra_image_base_address.png "Obtain image base address using Ghidra")
+
+From our previous analysis in "[Manual (Reversed) Code Review](#manual-reversed-code-review)" section, in the UnCrackable Level 1 application, the value of the hidden string is stored in a label with *hidden* flag set. In the disassembly, the text value of this label is stored in register `X21`, `mov`d from `X0`, at offset `0x100004520`. This is our first address, _breakpoint offset_.
```
```suggestion
From our previous analysis of the UnCrackable Level 1 application in "[Manual (Reversed) Code Review](#manual-reversed-code-review)" section, the value of the hidden string is stored in a label with the `hidden` flag set. In the disassembly, the text value of this label is stored in register `X21`, stored via `mov` from `X0`, at offset `0x100004520`. This is our _breakpoint offset_.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+![Breakpoint address using Ghidra](Images/Chapters/0x06c/debugging_ghidra_breakpoint.png "Breakpoint address using Ghidra")
+
+For the second address, we need to determine the ASLR shift offset for a given process. As mentioned before, iOS is a modern operating system with multiple techniques implemented to mitigate code execution attacks, one such technique being Address Space Randomization Layout (ASLR). On every new execution of an application, a random ASLR shift offset is generated, and various process' data structu ...
```
Some content moved above.

```suggestion
For the second address, we need to determine the _ASLR shift offset_ for a given process. The ASLR offset can be determined by using the lldb command `image list -o -f`. The output is shown in the screenshot below.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+Some groundwork is needed before setting a breakpoint using memory addresses. It requires determining two addresses:
+1. Breakpoint offset: The _address offset_ of the code where we want to set a breakpoint. This address is obtained by performing static analysis of the code in a disassembler like Ghidra.
+2. ASLR shift offset: The _ASLR shift offset_ for the current process. Since ASLR offset is randomly generated on every new instance of an application, this has to be obtained for every debugging session individually. This is determined using the debugger itself.
+
```
Content moved from below and made a note.

```suggestion

> iOS is a modern operating system with multiple techniques implemented to mitigate code execution attacks, one such technique being Address Space Randomization Layout (ASLR). On every new execution of an application, a random ASLR shift offset is generated, and various process' data structures are shifted by this offset. 

```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+In the output, the first column contains the sequence number of the image ([X]), the second column contains the randomly generated ASLR offset, while 3rd column contains the full path of the image and towards the end, content in the bracket shows the image base address after adding ASLR offset to the original image base address (`0x100000000` + `0x40000` = `0x100040000`). You will notice the imag ...
+
+In the above output, you may also notice that many of the full paths of the images do not point to a path on the iOS device, instead it points to a location on the machine on which `lldb` is running. These are system libraries for which debug symbols are available on the machine (part of Xcode iOS development toolkit) to aid in application development and debugging. For these libraries we can use ...
```
Very interesting note here ;) just a little adaption from my side + I made it a "side note" as the following paragraph actually continues what you're telling in the previous to this one. 

```suggestion
> In the above output, you may also notice that many of the paths listed as images do not point to the file system on the iOS device. Instead, they point to a certain location on the machine on which lldb is running. These images are system libraries for which debug symbols are available on the machine to aid in application development and debugging (as part of the Xcode iOS SDK). Therefore, you may set breakpoints to these libraries directly by using function names.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+In the above output, you may also notice that many of the full paths of the images do not point to a path on the iOS device, instead it points to a location on the machine on which `lldb` is running. These are system libraries for which debug symbols are available on the machine (part of Xcode iOS development toolkit) to aid in application development and debugging. For these libraries we can use ...
+
+Once the breakpoint is hit, the execution is halted, and we can access and explore the current state of the process. In this case we know from our static analysis that register `X0` contains the hidden string, thus lets explore it. In `lldb` we can print the Obj-C object using `po` command.
```
```suggestion
After putting the breakpoint and running the app, the execution will be halted once the breakpoint is hit. Now you can access and explore the current state of the process. In this case, you know from the previous static analysis that the register `X0` contains the hidden string, thus let's explore it. In lldb you can print Objective-C objects using the `po` (_print object_) command.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+![Setting breakpoint in lldb](Images/Chapters/0x06c/debugging_lldb_breakpoint_solution.png "Setting breakpoint in lldb")
+
+Voila, aided with static analysis and debugger the crackme can be easily solved. There are plethora of features implemented in `lldb`, including changing value of the registers, changing value in the process memory and even automating tasks using Python scripts.
```
```suggestion
Voila, the crackme can be easily solved aided by static analysis and a debugger. There are plethora of features implemented in lldb, including changing the value of the registers, changing values in the process memory and even [automating tasks using Python scripts](https://lldb.llvm.org/use/python.html "lldb - Python Scripting").
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+Voila, aided with static analysis and debugger the crackme can be easily solved. There are plethora of features implemented in `lldb`, including changing value of the registers, changing value in the process memory and even automating tasks using Python scripts.
+
+Officially Apple recommends use of `lldb` for debugging purposes, but `gdb` can also be used on iOS. The techniques discussed above is applicable while debugging using `gdb` as well, provided the `lldb` specific commands are changed to `gdb` commands.
```
Added link to "GDB to LLDB command map".

```suggestion
Officially Apple recommends use of lldb for debugging purposes, but gdb can be also used on iOS. The techniques discussed above are applicable while debugging using gdb as well, provided the lldb specific commands are [changed to gdb commands](https://lldb.llvm.org/use/map.html "GDB to LLDB command map").
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
````diff
@@ -404,6 +404,60 @@ for armv7.
+debugserver -x backboard *:1234 /Applications/MobileSMS.app/MobileSMS
+```
+
+Attach to an already running application
````
Is this an extra step after -x ? or does -x also attach? 
```suggestion
Attach to an already running application:
```

su-vikas replied:
-x is the flag to tell debugserver to launch application

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
````diff
@@ -404,6 +404,60 @@ for armv7.
+```shell
+debugserver *:1234 -a "MobileSMS"
+```
+
````
```suggestion
You may connect now to the iOS device from your host computer:
```

su-vikas commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -404,6 +404,60 @@ for armv7.
+
+![Process image list](Images/Chapters/0x06c/debugging_lldb_image_list.png "Process image list")
+
+In the output, the first column contains the sequence number of the image ([X]), the second column contains the randomly generated ASLR offset, while 3rd column contains the full path of the image and towards the end, content in the bracket shows the image base address after adding ASLR offset to the original image base address (`0x100000000` + `0x40000` = `0x100040000`). You will notice the imag ...
```
Actually in the screenshot the 2nd column in total is for ASLR offset for that respective image. With "ASLR offset"  I want to point to the whole column.  Does that make sense or not obvious enough?

cpholguera replied:
If I look at the image, I can only see that the value in the first column is the ASLR offset (`0x40000`), only for the first row. Being the absolute address at the end of the row in parentheses (`0x100040000`).

For the other rows it seems to show absolute addresses. E.g. (`0x1000dc000`).

But maybe I'm understanding it wrongly. Or at least, that's being the cause of my confusion ^^

su-vikas commented:
@cpholguera fixed LLDB, GDB and offset with backticks formatting.  Need clarification on the screenshot part, the rest of the suggestions are good :)

cpholguera commented:
I'll take some time to look at the changes carefully. Thanks @su-vikas !

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
````diff
@@ -404,6 +404,67 @@ for armv7.
+debugserver -x backboard *:1234 /Applications/MobileSMS.app/MobileSMS
+```
+
+Attach to an already running application:
````
new line so that the linter does not complain ;)
```suggestion
Attach to an already running application:

```

su-vikas commented:
@cpholguera have a re-look now. Fixed the image, I hope it is no more confusing now, and also removed the backticks from addresses. Let me know if things look stable now?
